File: Design-Patterns-Using-Kotlin/Singleton/SingletonDemo.kt
class Database {
    init {
        println("Database created")
    }

    fun addData(data: String?) {
        println(data)
    }
}

class Singleton private constructor() {
    private val db: Database

    init {
        instance = this
        db = Database()
    }

    fun addData(data: String?) {
        db.addData(data)
    }

    companion object {
        private var instance: Singleton? = null
        
        fun getInstance(): Singleton? {
            if (instance == null) {
                Singleton()
            }
            return instance
        }
    }
}

fun main() {
    val s1 = Singleton.getInstance()
    val s2 = Singleton.getInstance()

    println(s1)
    println(s2)

    s2!!.addData("Hello, world!")
}

/*
Database created
Singleton@7344699f
Singleton@7344699f
Hello, world!
 */

File: Design-Patterns-Using-JavaScript/Singleton/SingletonDemo.js
class Database {
    constructor() {
        console.log("Database created");
    }

    addData(data) {
        console.log(data);
    }
}

class Singleton {
    constructor() {
        if (!Singleton.instance) {
            Singleton.instance = this;
            this.db = new Database();
        }
        return Singleton.instance;
    }

    static getInstance() {
        if (!Singleton.instance) {
            new Singleton();
        }
        return Singleton.instance;
    }

    addData(data) {
        this.db.addData(data);
    }
}

// Client code
const s1 = Singleton.getInstance();
const s2 = Singleton.getInstance();
console.log(s1 === s2); // true, because it's the same instance
s1.addData("Hello, world!");